# Student Performance Analysis

## Objective

This notebook aims to analyze student behavior in learning management systems as a predictor of learning success. The study will employ data mining techniques with different feature sets, including LMS data, course characteristics, and student activities, to predict student performance.

### Goals

1. Determine which data sources yield the highest accuracy in predicting student performance.
2. Investigate whether students with high activity levels outperform those with lower activity levels.

## Import Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set(color_codes = True)
import warnings

warnings.filterwarnings('ignore')
%matplotlib inline

## Load Dataset

**Let us read the data from an excel file**, <br>
We will print top few rows to understand about the various data columns

In [2]:
# Load the Excel file into a DataFrame
file_path = '2022-23 CST4070 Applied Data Analytics - Tools, Practical Big Data Handling, Cloud Distribution Grad.xlsx'
df = pd.read_excel(file_path)

df.head()

UserID  Campus Quiz: Formative test w1 (Real)  \
0   user33   Dubai                              -   
1  user172   Dubai                              -   
2   user57  Hendon                              -   
3   user93   Dubai                              -   
4  user158   Dubai                              -   

  Quiz: Formative test w1 (Jan) (Real) Quiz: Formative test w2 (Real)  \
0                                    -                              -   
1                                    -                              -   
2                                    -                              -   
3                                    -                              -   
4                                    -                              -   

  Quiz: Formative test w2 Jan31 (Real) Quiz: Formative test w3 (Real)  \
0                                    -                              -   
1                                    -                              -   
2                                    9                              -   
3                                    -                              -   
4                                    -                              -   

  Quiz: Formative test w3 Jan 31 (Real) Quiz: Formative test w4 (Real)  \
0                                     -                              -   
1                                     -                              -   
2                                     6                              -   
3                                     -                              -   
4                                     -                              -   

  Quiz: Formative test w4 Feb. 14 (Real)  ...  \
0                                      -  ...   
1                                      -  ...   
2                                   5.33  ...   
3                                      -  ...   
4                                      -  ...   

  Assignment: Component 6 - Individual Submission link (Real).1  \
0                                                  -              
1                                                  -              
2                                                  -              
3                                                  -              
4                                                  -              

  Quiz: Formative test week 9 (Real)  \
0                                  -   
1                                  -   
2                                  -   
3                                  -   
4                                  -   

  Quiz: Formative test week 9 (March 21) (Real)  \
0                                             -   
1                                             -   
2                                             -   
3                                             -   
4                                             -   

  Assignment: Appendix for Dubai CW1 (Real) Quiz: Formative test w10 (Real)  \
0                                         -                               -   
1                                         -                               -   
2                                         -                               -   
3                                         -                               -   
4                                         -                               -   

  Assignment: Submit your predictions here (Real)  \
0                                               -   
1                                               -   
2                                               -   
3                                               -   
4                                               -   

  Assignment: Submit your notebook here (Real)  \
0                                            -   
1                                            -   
2                                            -   
3                                            -   
4                                            -   

  Assignment: Compone

In [3]:
df.shape

(174, 75)

In [4]:
df.UserID.nunique()

174

The dataset contains multiple columns related to quizzes, assignments, and other components of the course with a total of 174 rows. Each row represents a student's performance in these activities.

In [5]:
# Filtering records where the campus is 'Hendon'
df_hendon = df[df['Campus'] == "Hendon"]

# Displaying the first few rows of the filtered DataFrame
df_hendon.head()

UserID  Campus Quiz: Formative test w1 (Real)  \
2    user57  Hendon                              -   
6    user61  Hendon                              -   
8   user111  Hendon                              -   
10   user97  Hendon                           3.33   
12  user120  Hendon                           6.67   

   Quiz: Formative test w1 (Jan) (Real) Quiz: Formative test w2 (Real)  \
2                                     -                              -   
6                                     -                              -   
8                                     -                              -   
10                                    -                              5   
12                                    -                           4.33   

   Quiz: Formative test w2 Jan31 (Real) Quiz: Formative test w3 (Real)  \
2                                     9                              -   
6                                     2                              -   
8                                     6                              -   
10                                    -                              7   
12                                    -                              3   

   Quiz: Formative test w3 Jan 31 (Real) Quiz: Formative test w4 (Real)  \
2                                      6                              -   
6                                      9                              -   
8                                      4                              -   
10                                     -                           3.33   
12                                     -                              -   

   Quiz: Formative test w4 Feb. 14 (Real)  ...  \
2                                    5.33  ...   
6                                    4.33  ...   
8                                       1  ...   
10                                      -  ...   
12                                      -  ...   

   Assignment: Component 6 - Individual Submission link (Real).1  \
2                                                   -              
6                                                   -              
8                                                   -              
10                                                  -              
12                                                  -              

   Quiz: Formative test week 9 (Real)  \
2                                   -   
6                                   -   
8                                   -   
10                                  -   
12                                  -   

   Quiz: Formative test week 9 (March 21) (Real)  \
2                                              -   
6                                              8   
8                                              -   
10                                             -   
12                                             -   

   Assignment: Appendix for Dubai CW1 (Real) Quiz: Formative test w10 (Real)  \
2                                          -                               -   
6                                          -                               -   
8                                          -                               -   
10                                         -                            2.67   
12                                         -                               -   

   Assignment: Submit your predictions here (Real)  \
2                                                -   
6                                                -   
8                                                -   
10                                               -   
12                                               -   

   Assignment: Submit your notebook here (Real)  \
2                                             -   
6                                             -   
8                                             -   
10                                            -   
12      

In [6]:
df_hendon.shape

(116, 75)

The filtered dataset contains multiple columns related to quizzes, assignments, and other components of the course, but only for students at the Hendon campus with a total of 116 rows

In [7]:
# File path for the second dataset
file_path_logs = 'logs.csv'

# Loading the second dataset into a DataFrame
df_logs = pd.read_csv(file_path_logs)

# Displaying the first few rows of the DataFrame
df_logs.head()

Time  UserID                                      Event context  \
0  18/06/23, 13:41  user25  Course: 2022-23 CST4070 Applied Data Analytics...   
1  18/06/23, 13:38  user25  Course: 2022-23 CST4070 Applied Data Analytics...   
2  18/06/23, 02:02  user58  Course: 2022-23 CST4070 Applied Data Analytics...   
3  17/06/23, 20:17  user18  Course: 2022-23 CST4070 Applied Data Analytics...   
4  17/06/23, 20:07  user18  Course: 2022-23 CST4070 Applied Data Analytics...   

  Component     Event name                                        Description  
0    System  Course viewed  The user with id '138154' viewed the course wi...  
1    System  Course viewed  The user with id '138154' viewed the course wi...  
2    System  Course viewed  The user with id '156217' viewed the course wi...  
3    System  Course viewed  The user with id '141836' viewed the course wi...  
4    System  Course viewed  The user with id '141836' viewed the course wi...

In [8]:
df_logs.shape

(113725, 6)

The logs dataset contains information about user activities, including the time of the activity, the user ID, the event context, the component involved, the event name, and a description of the event. The loag dataset contains a record of 113,725 records

In [9]:
# Joining the two DataFrames using the 'UserID' column
df_joined = pd.merge(left=df_hendon, right=df_logs, left_on='UserID', right_on='UserID')

# Displaying the first few rows of the joined DataFrame
df_joined.head()

UserID  Campus Quiz: Formative test w1 (Real)  \
0  user57  Hendon                              -   
1  user57  Hendon                              -   
2  user57  Hendon                              -   
3  user57  Hendon                              -   
4  user57  Hendon                              -   

  Quiz: Formative test w1 (Jan) (Real) Quiz: Formative test w2 (Real)  \
0                                    -                              -   
1                                    -                              -   
2                                    -                              -   
3                                    -                              -   
4                                    -                              -   

  Quiz: Formative test w2 Jan31 (Real) Quiz: Formative test w3 (Real)  \
0                                    9                              -   
1                                    9                              -   
2                                    9                              -   
3                                    9                              -   
4                                    9                              -   

  Quiz: Formative test w3 Jan 31 (Real) Quiz: Formative test w4 (Real)  \
0                                     6                              -   
1                                     6                              -   
2                                     6                              -   
3                                     6                              -   
4                                     6                              -   

  Quiz: Formative test w4 Feb. 14 (Real)  ...  \
0                                   5.33  ...   
1                                   5.33  ...   
2                                   5.33  ...   
3                                   5.33  ...   
4                                   5.33  ...   

  Assignment: Submit your predictions here (Real)  \
0                                               -   
1                                               -   
2                                               -   
3                                               -   
4                                               -   

  Assignment: Submit your notebook here (Real)  \
0                                            -   
1                                            -   
2                                            -   
3                                            -   
4                                            -   

  Assignment: Component 6 - Group PDF Submission link (Real)  \
0                                                  -           
1                                                  -           
2                                                  -           
3                                                  -           
4                                                  -           

  Assignment: Component 6 - Group Screenshot Submission link (Real)  \
0                                                  -                  
1                                                  -                  
2                                                  -                  
3                                                  -                  
4                                                  -                  

  Course total (Real)             Time  \
0               61.13  13/06/23, 16:08   
1               61.13  13/06/23, 16:08   
2               61.13  13/06/23, 16:07   
3               61.13  13/06/23, 15:23   
4               61.13  13/06/23, 15:22   

                                       Event context Component  \
0  Course: 2022-23 CST4070 Applied Data Analytics...    System   
1                             Folder: Lecture slides    Folder   
2  Course: 2022-23 CST4070 Applied Data Analytics...    System   
3                             Folder: Lecture slides    Folder   
4  Course: 2022-23 CST4070 Applied Data Analytics...    S

In [10]:
df_joined.shape

(100058, 80)

In [11]:
df_joined.UserID.nunique()

116

The joined DataFrame contains columns from both the original DataFrames, allowing us to analyze student performance in relation to their activity logs. It contains 100,058 records

## Data Cleaning

Before diving into the analysis, it's crucial to clean the data. This involves:

1. Removing duplicates
2. Handling missing values
3. Converting data types, if necessary
4. Renaming columns for better readability

Let's start by examining the missing values in both datasets.

### Removing Duplicates


In [12]:
# Checking for duplicate records in the DataFrame
duplicates = df_joined.duplicated().sum()
duplicates

16149

It appears that there are 16,149 duplicate records in the joined DataFrame. These duplicates need to be removed to ensure the quality of the data.

In [13]:
# Removing duplicate records from the DataFrame
df_joined_clean = df_joined.drop_duplicates()

# Verifying that duplicates have been removed
remaining_duplicates = df_joined_clean.duplicated().sum()
remaining_duplicates

0

All 16,149 duplicate records have been successfully removed, leaving us with a clean dataset

### Handling Missing Values

In [14]:
# Checking for missing values in the DataFrame
missing_values = df_joined.isnull().sum()
missing_values[missing_values > 0]

Series([], dtype: int64)

It appears that there are no missing values in the joined DataFrame.

### Replacing Hyphens with Spaces

In [15]:
# Identifying columns that are of 'object' data type
object_columns = df_joined_clean.select_dtypes(include=['object']).columns

# Checking if these columns contain '-' and replacing them with average of the column
for col in object_columns:
    df_joined_clean[col] = df_joined_clean.loc[:,[col]].replace('-','')
    
# Verifying the changes
df_joined_clean.head()

UserID  Campus Quiz: Formative test w1 (Real)  \
0  user57  Hendon                                  
1  user57  Hendon                                  
2  user57  Hendon                                  
3  user57  Hendon                                  
4  user57  Hendon                                  

  Quiz: Formative test w1 (Jan) (Real) Quiz: Formative test w2 (Real)  \
0                                                                       
1                                                                       
2                                                                       
3                                                                       
4                                                                       

  Quiz: Formative test w2 Jan31 (Real) Quiz: Formative test w3 (Real)  \
0                                    9                                  
1                                    9                                  
2                                    9                                  
3                                    9                                  
4                                    9                                  

  Quiz: Formative test w3 Jan 31 (Real) Quiz: Formative test w4 (Real)  \
0                                     6                                  
1                                     6                                  
2                                     6                                  
3                                     6                                  
4                                     6                                  

  Quiz: Formative test w4 Feb. 14 (Real)  ...  \
0                                   5.33  ...   
1                                   5.33  ...   
2                                   5.33  ...   
3                                   5.33  ...   
4                                   5.33  ...   

  Assignment: Submit your predictions here (Real)  \
0                                                   
1                                                   
2                                                   
3                                                   
4                                                   

  Assignment: Submit your notebook here (Real)  \
0                                                
1                                                
2                                                
3                                                
4                                                

  Assignment: Component 6 - Group PDF Submission link (Real)  \
0                                                              
1                                                              
2                                                              
3                                                              
4                                                              

  Assignment: Component 6 - Group Screenshot Submission link (Real)  \
0                                                                     
1                                                                     
2                                                                     
3                                                                     
4                                                                     

  Course total (Real)             Time  \
0               61.13  13/06/23, 16:08   
1               61.13  13/06/23, 16:08   
2               61.13  13/06/23, 16:07   
3               61.13  13/06/23, 15:23   
4               61.13  13/06/23, 15:22   

                                       Event context Component  \
0  Course: 2022-23 CST4070 Applied Data Analytics...    System   
1                             Folder: Lecture slides    Folder   
2  Course: 2022-23 CST4070 Applied Data Analytics...    System   
3                             Folder: Lecture slides    Folder   
4  Course: 2022-23 CST4070 Applied Data Analytics...    S

### Data Type Conversion

In [16]:
# Checking the data types of each column in the DataFrame
df_joined_clean.dtypes

UserID                                  object
Campus                                  object
Quiz: Formative test w1 (Real)          object
Quiz: Formative test w1 (Jan) (Real)    object
Quiz: Formative test w2 (Real)          object
                                         ...  
Time                                    object
Event context                           object
Component                               object
Event name                              object
Description                             object
Length: 80, dtype: object

In [18]:
# Identifying other columns that should be numerical but are currently 'object'
object_columns = df_joined_clean.columns[2:]

# Attempting to convert these columns to numerical data types
for col in object_columns:
    try:
        if col == "Time":
            continue
            
        df_joined_clean[col] = pd.to_numeric(df_joined_clean[col], errors='coerce')
    except Exception as e:
        print(f'Could not convert column {col}: {e}')

# Checking the updated data types
updated_data_types = df_joined_clean.dtypes
updated_data_types

UserID                                   object
Campus                                   object
Quiz: Formative test w1 (Real)          float64
Quiz: Formative test w1 (Jan) (Real)    float64
Quiz: Formative test w2 (Real)          float64
                                         ...   
Time                                     object
Event context                           float64
Component                               float64
Event name                              float64
Description                             float64
Length: 80, dtype: object

In [20]:
# Convert Time Column to DateTime
df_joined_clean['Time'] =  pd.to_datetime(df_joined_clean['Time'], infer_datetime_format=True)

df_joined_clean['Time'].dtype

dtype('<M8[ns]')

In [21]:
df_joined_clean.head()

UserID  Campus  Quiz: Formative test w1 (Real)  \
0  user57  Hendon                             NaN   
1  user57  Hendon                             NaN   
2  user57  Hendon                             NaN   
3  user57  Hendon                             NaN   
4  user57  Hendon                             NaN   

   Quiz: Formative test w1 (Jan) (Real)  Quiz: Formative test w2 (Real)  \
0                                   NaN                             NaN   
1                                   NaN                             NaN   
2                                   NaN                             NaN   
3                                   NaN                             NaN   
4                                   NaN                             NaN   

   Quiz: Formative test w2 Jan31 (Real)  Quiz: Formative test w3 (Real)  \
0                                   9.0                             NaN   
1                                   9.0                             NaN   
2                                   9.0                             NaN   
3                                   9.0                             NaN   
4                                   9.0                             NaN   

   Quiz: Formative test w3 Jan 31 (Real)  Quiz: Formative test w4 (Real)  \
0                                    6.0                             NaN   
1                                    6.0                             NaN   
2                                    6.0                             NaN   
3                                    6.0                             NaN   
4                                    6.0                             NaN   

   Quiz: Formative test w4 Feb. 14 (Real)  ...  \
0                                    5.33  ...   
1                                    5.33  ...   
2                                    5.33  ...   
3                                    5.33  ...   
4                                    5.33  ...   

   Assignment: Submit your predictions here (Real)  \
0                                              NaN   
1                                              NaN   
2                                              NaN   
3                                              NaN   
4                                              NaN   

   Assignment: Submit your notebook here (Real)  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   

   Assignment: Component 6 - Group PDF Submission link (Real)  \
0                                                NaN            
1                                                NaN            
2                                                NaN            
3                                                NaN            
4                                                NaN            

   Assignment: Component 6 - Group Screenshot Submission link (Real)  \
0                                                NaN                   
1                                                NaN                   
2                                                NaN                   
3                                                NaN                   
4                                                NaN                   

   Course total (Real)                Time  Event context  Component  \
0                61.13 2023-06-13 16:08:00            NaN        NaN   
1                61.13 2023-06-13 16:08:00            NaN        NaN   
2                61.13 2023-06-13 16:07:00            NaN        NaN   
3                61.13 2023-06-13 15:23:00            NaN        NaN   
4                61.13 2023-06-13 15:22:00            NaN        NaN   

   Event name  Description  
0         NaN          NaN  
1         NaN          NaN  
2         NaN          NaN  
3         NaN      

### Handling Missing Values in Numeric Columns

In [25]:
# Identify numeric columns with missing values
numeric_columns_with_missing_values = df_joined_clean.select_dtypes(include=['number']).isnull().sum()
numeric_columns_with_missing_values = numeric_columns_with_missing_values[numeric_columns_with_missing_values > 0]
numeric_columns_with_missing_values

Quiz: Formative test w1 (Real)                                       35887
Quiz: Formative test w1 (Jan) (Real)                                 83909
Quiz: Formative test w2 (Real)                                       25755
Quiz: Formative test w2 Jan31 (Real)                                 65875
Quiz: Formative test w3 (Real)                                       23399
                                                                     ...  
Assignment: Component 6 - Group Screenshot Submission link (Real)    83909
Event context                                                        83909
Component                                                            83909
Event name                                                           83909
Description                                                          83909
Length: 76, dtype: int64

I have identified several numeric columns with missing values. For example:

- `Quiz: Formative test w1 (Real)`: 35,887 missing values
- `Quiz: Formative test w1 (Jan) (Real)`: 83,909 missing values
- `Quiz: Formative test w2 (Real)`: 25,755 missing values

In [26]:
# Perform mean imputation for numeric columns with missing values
numeric_columns = numeric_columns_with_missing_values.index
df_joined_clean[numeric_columns] = df_joined_clean[numeric_columns].fillna(df_joined_clean[numeric_columns].mean())

# Verify if there are any missing values left in these columns
missing_values_after_imputation = df_joined_clean[numeric_columns].isnull().sum()
missing_values_after_imputation

Quiz: Formative test w1 (Real)                                           0
Quiz: Formative test w1 (Jan) (Real)                                 83909
Quiz: Formative test w2 (Real)                                           0
Quiz: Formative test w2 Jan31 (Real)                                     0
Quiz: Formative test w3 (Real)                                           0
                                                                     ...  
Assignment: Component 6 - Group Screenshot Submission link (Real)    83909
Event context                                                        83909
Component                                                            83909
Event name                                                           83909
Description                                                          83909
Length: 76, dtype: int64

After performing mean imputation on the numeric columns, I find out that most of the missing values have been filled. However, there are still some columns with missing values, such as Quiz: Formative test w1 (Jan) (Real) and Turnitin Assignment 2: Component 6 Dubai Campus Group Submission (Real).

The reason these columns still have missing values is that they were entirely made up of NaNs, so there was no mean value to impute. These columns do not contribute any information and can be safely removed to simplify the dataset.

In [28]:
# Identify columns that are entirely made up of NaNs
columns_with_all_nans = df_joined_clean.columns[df_joined_clean.isna().all()].tolist()

# Remove these columns from the DataFrame
df_joined_clean.drop(columns=columns_with_all_nans, inplace=True)

# Verify if these columns have been removed
remaining_columns = df_joined_clean.columns.tolist()
remaining_columns

['UserID',
 'Campus',
 'Quiz: Formative test w1 (Real)',
 'Quiz: Formative test w2 (Real)',
 'Quiz: Formative test w2 Jan31 (Real)',
 'Quiz: Formative test w3 (Real)',
 'Quiz: Formative test w3 Jan 31 (Real)',
 'Quiz: Formative test w4 (Real)',
 'Quiz: Formative test w4 Feb. 14 (Real)',
 'Turnitin Assignment 2: Component 6 - Group Submission link (Real)',
 'Turnitin Assignment 2: Component 1 - Submission link (Hendon) (Real)',
 'Turnitin Assignment 2: Component 1 - Submission link (Hendon Campus) (Real)',
 'Quiz: Formative test w7 (Real)',
 'Quiz: Component 2 - In-class test (Real)',
 'Quiz: Component 2 - In-class test (resit) (Real)',
 'Quiz: Component 2 - In-class test (Hendon Campus) (Real)',
 'Quiz: Formative test w13 (Real)',
 'Quiz: Formative test w14 (2021) (Real)',
 'Quiz: Formative test w15 (Real)',
 'Quiz: Formative test w16 (Real)',
 'Quiz: Formative test w17 (Real).1',
 'Quiz: Component 4 - in-class test (Real)',
 'Quiz: Component 4 - in-class test (resit) (Real)',
 'Quiz: 

In [31]:
# Check for any remaining missing values in the dataset
remaining_missing_values = df_joined_clean.isnull().sum().sum()
remaining_missing_values

0

After checking for any remaining missing values, we found that the dataset is now entirely clean with **0 missing values**.

This means the dataset is ready for further analysis and exploration.